dotnet/src/SemanticKernel.Abstractions/AI/PromptExecutionSettings.cs

```diff
@@ -39,5 +52,43 @@ public class PromptExecutionSettings
     /// Avoid using this property if possible. Instead, use one of the classes that extends <see cref="PromptExecutionSettings"/>.
     /// </remarks>
     [JsonExtensionData]
-    public Dictionary<string, object>? ExtensionData { get; set; }
+    public IDictionary<string, object>? ExtensionData
+    {
+        get => this._extensionData;
+
+        set
+        {
+            if (this._isFrozen)
+            {
+                throw new InvalidOperationException("PromptExecutionSettings are frozen and ExtensionData cannot be modified.");
+            }
+            this._extensionData = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets a value that indicates whether the <see cref="PromptExecutionSettings"/> are currently modifiable.
+    /// </summary>
+    public bool IsFrozen
+    {
+        get => this._isFrozen;
+    }
+
+    /// <summary>
+    /// Makes the current <see cref="PromptExecutionSettings"/> unmodifiable and sets its IsFrozen property to true.
+    /// </summary>
+    public PromptExecutionSettings Freeze()
+    {
+        this._isFrozen = true;
+
+        return this;
```

Comment:
Is returning `this` necessary? Having this method return a `PromptExecutionSettings` makes it seem to me like it's returning a new frozen copy rather than mutating itself.

That would be another approach that could be taken if desired, e.g. a Clone method and a Freeze method, or alternatively a method that creates a read-only clone, rather than mutating this.

Reply:
I've added a `Clone()` method and kept `Freeze()` but happy to combine these.